datastream-server/src/main/java/com/linkedin/datastream/server/EventProducer.java

```diff
@@ -272,20 +273,28 @@ private void performSlaRelatedLogging(DatastreamRecordMetadata metadata, long ev
     }
 
     if (_numEventsOutsideAltSlaLogEnabled) {
-      if (sourceToDestinationLatencyMs > _availabilityThresholdAlternateSlaMs) {
-        TopicPartition topicPartition = new TopicPartition(metadata.getTopic(), metadata.getSourcePartition());
-        int numEvents = _trackEventsOutsideAltSlaMap.getOrDefault(topicPartition, 0);
-        _trackEventsOutsideAltSlaMap.put(topicPartition, numEvents + 1);
-      }
+      try {
+        if (sourceToDestinationLatencyMs > _availabilityThresholdAlternateSlaMs) {
+          TopicPartition topicPartition = new TopicPartition(metadata.getTopic(), metadata.getSourcePartition());
+          int numEvents = _trackEventsOutsideAltSlaMap.getOrDefault(topicPartition, 0);
+          _trackEventsOutsideAltSlaMap.put(topicPartition, numEvents + 1);
+        }
 
-      long timeSinceLastLog = System.currentTimeMillis() - _lastEventsOutsideAltSlaLogTimeMs;
-      if (timeSinceLastLog >= _numEventsOutsideAltSlaFrequencyMs) {
-        _trackEventsOutsideAltSlaMap.forEach((topicPartition, numEvents) ->
-            _logger.warn("{} had {} event(s) with latency greater than alternate SLA of {} ms in the last {} ms for "
-                    + "datastream {}", topicPartition, numEvents, _availabilityThresholdAlternateSlaMs,
-                timeSinceLastLog, getDatastreamName()));
-        _trackEventsOutsideAltSlaMap.clear();
-        _lastEventsOutsideAltSlaLogTimeMs = System.currentTimeMillis();
+        long timeSinceLastLog = System.currentTimeMillis() - _lastEventsOutsideAltSlaLogTimeMs;
+        if (timeSinceLastLog >= _numEventsOutsideAltSlaFrequencyMs) {
+          _trackEventsOutsideAltSlaMap.forEach((topicPartition, numEvents) -> _logger.warn(
+              "{} had {} event(s) with latency greater than alternate SLA of {} ms in the last {} ms for "
+                  + "datastream {}", topicPartition, numEvents, _availabilityThresholdAlternateSlaMs, timeSinceLastLog,
+              getDatastreamName()));
+          _trackEventsOutsideAltSlaMap.clear();
+          _lastEventsOutsideAltSlaLogTimeMs = System.currentTimeMillis();
+        }
+      } catch (ConcurrentModificationException | NullPointerException | IllegalArgumentException | ClassCastException
```

Comment:
ConcurrentHashMap will not throw ConcurrentModificationException. You can get rid of it.

Reply:
done